course-content/21_getting_data_with_pandas.qmd

```diff
@@ -0,0 +1,427 @@
+series_mixed
+```
+
+When pandas sees data that cannot be obviously converted into a single datatype, it leaves each individual entry alone, as whatever its original datatype happened to be. Be careful when you see this! Many operations on pandas dataframes apply to the entire column (similar to how a new excel column is often created with a function that applies to a previous column). If this operation is built to only work on a single datatype, we might run into errors! To avoid this, we can utilize the `.astype()` methods. Possible arguments for `astype()` include the standard datatypes:
+
+ * `int`
+ * `float`
+ * `str`
+ * `bool`
+ 
+However if you need to be more specific, you can be as well. These arguments are all required to be included in quotations, as they refer to aliases for pandas specific datatypes:
+
+ * `'Int8'`, `'Int32'`, `'Int64'` provide nullable integer types (note the capital I!)
+ * `'string'` provides access to a series that will be string specific, instead of the general `objects` type. Pandas recommends using this type for strings whenever possible. 
+
+```{python}
+display(series_y)
+display(series_y.astype(int))
+display(series_y.astype(bool))
+display(series_y.astype(str))
+display(series_y.astype('string'))
+```
+
+You will have noticed that each of these series has an index associated with it. We can access series indicies using the `index` attribute:
+
+```{python}
+series_a = pd.Series([2, 4, 6])
+display(series_a)
+display(series_a.index)
+```
+
+When an index is first assigned to a series, it is automatically assigned as an integer index, with similar properties to a list index (starts at 0, can be sliced, etc.). However we can change this index to be whatever we want by directly modifying the index attribute:
+
+```{python}
+series_a.index = ['a', 'b', 'c']
+display(series_a.index)
+display(series_a)
+```
+
+A useful feature in pandas is the ability to `reindex` the dataset. Reindexing a dataset will do two things:
+
+1. Reorder the data according to the order we ask for. 
+2. Add new rows for indicies that are missing in the original dataset but are included in our new index. 
+
+One popular use for this method is in filling out a timeseries dataset: if there were missing years in a dataset but we do not wish to simply skip over them, we can add the extra years to the index. 
+
+```{python}
+series_a = series_a.reindex(['d', 'c', 'b', 'a'])
+display(series_a.index)
+display(series_a)
+```
+
+:::{.callout-tip}
+## Beware NaN values!
+
+In python, `NaN` is considered a `float` value. In the above example, we introduced a missing value for the new 'd' index, which defaults to the `NaN` float value. Because its type is float, this converted our entire Series to float as well. If we wish to keep the series as a float, we can coerce it back using the `astype()` method with one of the nullable integer types. This will introduce a slightly different looking null value that works for integers! 
```

Comment:
Series vs series, I am struggling to figure out a consistent rule as to when to capitalize in text.

Reply:
Good question. I think on the whole it's fairly safe to go lowercase, but I'm probably not being consistent myself.